multicast/src/main/kotlin/com/dropbox/flow/multicast/ChannelManager.kt

```diff
@@ -75,12 +77,24 @@ internal class ChannelManager<T>(
         when (msg) {
             is Message.AddChannel -> doAdd(msg)
             is Message.RemoveChannel -> doRemove(msg.channel)
-            is Message.DispatchValue -> doDispatchValue(msg)
-            is Message.DispatchError -> doDispatchError(msg)
-            is Message.UpstreamFinished -> doHandleUpstreamClose(msg.producer)
+            is Message.Dispatch.Value -> doDispatchValue(msg)
+            is Message.Dispatch.Error -> doDispatchError(msg)
+            is Message.Dispatch.UpstreamFinished -> doHandleUpstreamClose(msg.producer)
         }
     }
 
+    suspend fun addDownstream(channel: SendChannel<Message.Dispatch.Value<T>>) =
```

Comment:
after reading rest of the code, i'm happy w/ this version instead of the explicit message passing.